File: leetcode/301-400/317-shortest-distance-from-all-buildings/bench/shortest_distance.py
"""Benchmark lane for LeetCode 317 — Python mirror of bench/shortest_distance.kara.

Build the grid once (20% obstacles, BUILDINGS buildings on corner-reachable
empty cells), then PASSES one-BFS-per-building passes, each after relocating one
building to an empty cell chosen from the checksum (moved back after).
"""

ROWS = 360
COLS = 360
BUILDINGS = 20
OBSTACLE_PCT = 20
PASSES = 30
MASK = 1073741823


def lcg(s):
    return (s * 1103515245 + 12345) & 0x7FFFFFFF


def shortest_distance(grid, rows, cols):
    n = rows * cols
    total = [0] * n
    reach = [0] * n
    seen = [0] * n
    dist = [0] * n
    b = 0
    for src in range(n):
        if grid[src] != 1:
            continue
        b += 1
        seen[src] = b
        dist[src] = 0
        q = [src]
        head = 0
        while head < len(q):
            cell = q[head]
            head += 1
            r, c = divmod(cell, cols)
            d = dist[cell] + 1
            if r > 0:
                nb = cell - cols
                if grid[nb] == 0 and seen[nb] != b:
                    seen[nb] = b; dist[nb] = d; total[nb] += d; reach[nb] += 1; q.append(nb)
            if r < rows - 1:
                nb = cell + cols
                if grid[nb] == 0 and seen[nb] != b:
                    seen[nb] = b; dist[nb] = d; total[nb] += d; reach[nb] += 1; q.append(nb)
            if c > 0:
                nb = cell - 1
                if grid[nb] == 0 and seen[nb] != b:
                    seen[nb] = b; dist[nb] = d; total[nb] += d; reach[nb] += 1; q.append(nb)
            if c < cols - 1:
                nb = cell + 1
                if grid[nb] == 0 and seen[nb] != b:
                    seen[nb] = b; dist[nb] = d; total[nb] += d; reach[nb] += 1; q.append(nb)
    best = -1
    for i in range(n):
        if grid[i] == 0 and reach[i] == b and (best < 0 or total[i] < best):
            best = total[i]
    return best


def main():
    n = ROWS * COLS
    seed = 317
    grid = [0] * n
    for i in range(n):
        seed = lcg(seed)
        grid[i] = 2 if (seed // 65536) % 100 < OBSTACLE_PCT else 0
    grid[0] = 0

    reachable = [False] * n
    q = [0]
    reachable[0] = True
    head = 0
    while head < len(q):
        cell = q[head]
        head += 1
        r, c = divmod(cell, COLS)
        if r > 0 and grid[cell - COLS] != 2 and not reachable[cell - COLS]:
            reachable[cell - COLS] = True; q.append(cell - COLS)
        if r < ROWS - 1 and grid[cell + COLS] != 2 and not reachable[cell + COLS]:
            reachable[cell + COLS] = True; q.append(cell + COLS)
        if c > 0 and grid[cell - 1] != 2 and not reachable[cell - 1]:
            reachable[cell - 1] = True; q.append(cell - 1)
        if c < COLS - 1 and grid[cell + 1] != 2 and not reachable[cell + 1]:
            reachable[cell + 1] = True; q.append(cell + 1)

    sites = []
    while len(sites) < BUILDINGS:
        seed = lcg(seed)
        p = (seed // 256) % n
        if grid[p] == 0 and reachable[p]:
            grid[p] = 1
            sites.append(p)

    checksum = 0
    for pass_ in range(PASSES):
        old = sites[pass_ % BUILDINGS]
        i = checksum % n
        while grid[i] != 0:
            i = (i + 1) % n
        grid[old] = 0
        grid[i] = 1
        ans = shortest_distance(grid, ROWS, COLS)
        checksum = (checksum * 31 + ans + 7) & MASK
        grid[i] = 0
        grid[old] = 1
    print(f"checksum {checksum}")


if __name__ == "__main__":
    main()

File: leetcode/301-400/317-shortest-distance-from-all-buildings/bench/shortest_distance.rs
// Benchmark lane for LeetCode 317 — Rust mirror of bench/shortest_distance.kara.
// Build the grid once (20% obstacles, BUILDINGS buildings on corner-reachable
// empty cells), then PASSES one-BFS-per-building passes, each after relocating
// one building to an empty cell chosen from the checksum (moved back after).

const ROWS: i64 = 360;
const COLS: i64 = 360;
const BUILDINGS: i64 = 20;
const OBSTACLE_PCT: i64 = 20;
const PASSES: i64 = 30;
const MASK: i64 = 1073741823;

fn lcg(s: i64) -> i64 {
    (s * 1103515245 + 12345) & 0x7fffffff
}

fn shortest_distance(grid: &[i64], rows: i64, cols: i64) -> i64 {
    let n = (rows * cols) as usize;
    let mut total = vec![0i64; n];
    let mut reach = vec![0i64; n];
    let mut seen = vec![0i64; n];
    let mut dist = vec![0i64; n];
    let mut b: i64 = 0;
    for src in 0..n {
        if grid[src] != 1 {
            continue;
        }
        b += 1;
        seen[src] = b;
        dist[src] = 0;
        let mut q: Vec<usize> = Vec::new();
        q.push(src);
        let mut head = 0;
        while head < q.len() {
            let cell = q[head];
            head += 1;
            let r = cell as i64 / cols;
            let c = cell as i64 % cols;
            let d = dist[cell] + 1;
            let mut visit = |nb: usize, q: &mut Vec<usize>| {
                if grid[nb] == 0 && seen[nb] != b {
                    seen[nb] = b;
                    dist[nb] = d;
                    total[nb] += d;
                    reach[nb] += 1;
                    q.push(nb);
                }
            };
            if r > 0 { visit(cell - cols as usize, &mut q); }
            if r < rows - 1 { visit(cell + cols as usize, &mut q); }
            if c > 0 { visit(cell - 1, &mut q); }
            if c < cols - 1 { visit(cell + 1, &mut q); }
        }
    }
    let mut best: i64 = -1;
    for i in 0..n {
        if grid[i] == 0 && reach[i] == b && (best < 0 || total[i] < best) {
            best = total[i];
        }
    }
    best
}

fn main() {
    let n = (ROWS * COLS) as usize;
    let mut seed: i64 = 317;
    let mut grid: Vec<i64> = Vec::with_capacity(n);
    for _ in 0..n {
        seed = lcg(seed);
        grid.push(if (seed / 65536) % 100 < OBSTACLE_PCT { 2 } else { 0 });
    }
    grid[0] = 0;

    let mut reachable = vec![false; n];
    let mut q: Vec<usize> = Vec::new();
    reachable[0] = true;
    q.push(0);
    let mut head = 0;
    while head < q.len() {
        let cell = q[head];
        head += 1;
        let r = cell as i64 / COLS;
        let c = cell as i64 % COLS;
        let cols = COLS as usize;
        if r > 0 && grid[cell - cols] != 2 && !reachable[cell - cols] { reachable[cell - cols] = true; q.push(cell - cols); }
        if r < ROWS - 1 && grid[cell + cols] != 2 && !reachable[cell + cols] { reachable[cell + cols] = true; q.push(cell + cols); }
        if c > 0 && grid[cell - 1] != 2 && !reachable[cell - 1] { reachable[cell - 1] = true; q.push(cell - 1); }
        if c < COLS - 1 && grid[cell + 1] != 2 && !reachable[cell + 1] { reachable[cell + 1] = true; q.push(cell + 1); }
    }

    let mut sites: Vec<usize> = Vec::new();
    while (sites.len() as i64) < BUILDINGS {
        seed = lcg(seed);
        let p = ((seed / 256) % n as i64) as usize;
        if grid[p] == 0 && reachable[p] {
            grid[p] = 1;
            sites.push(p);
        }
    }

    let mut checksum: i64 = 0;
    for pass in 0..PASSES {
        let old = sites[(pass % BUILDINGS) as usize];
        let mut i = (checksum % n as i64) as usize;
        while grid[i] != 0 {
            i = (i + 1) % n;
        }
        grid[old] = 0;
        grid[i] = 1;
        let ans = shortest_distance(&grid, ROWS, COLS);
        checksum = (checksum * 31 + ans + 7) & MASK;
        grid[i] = 0;
        grid[old] = 1;
    }
    println!("checksum {}", checksum);
}
